File: backend/app/api/players.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.constants import SKILL_FIELDS
from app.database import get_db
from app.models.player import Player
from app.models.player_skill_history import PlayerSkillHistory

router = APIRouter()

PLAYER_DETAIL_FIELDS = [
    "id", "first_name", "last_name", "age", "age_days", "tsi", "form",
    "stamina", "speed", "injury_days", "salary",
    "goalkeeper", "defending", "playmaking", "winger", "passing", "scoring", "set_pieces",
    "leadership", "experience", "loyalty", "market_value", "speciality",
    "last_match_rating", "transfer_listed", "country_id", "homegrown", "data_source",
]


@router.get("/players/{player_id}")
def get_player(player_id: int, db: Session = Depends(get_db)):
    p = db.get(Player, player_id)
    if not p:
        raise HTTPException(status_code=404, detail="Player not found")
    return {field: getattr(p, field) for field in PLAYER_DETAIL_FIELDS}


@router.get("/players/{player_id}/history")
def get_player_history(player_id: int, db: Session = Depends(get_db)):
    p = db.get(Player, player_id)
    if not p:
        raise HTTPException(status_code=404, detail="Player not found")
    rows = (
        db.query(PlayerSkillHistory)
        .filter_by(player_id=player_id)
        .order_by(PlayerSkillHistory.snapshot_date)
        .all()
    )
    return [
        {"snapshot_date": r.snapshot_date.isoformat(),
         **{f: getattr(r, f) for f in SKILL_FIELDS}}
        for r in rows
    ]

File: frontend/src/api/players.ts
import { apiFetch } from './client'

export interface PlayerDetail {
  id: number
  first_name: string
  last_name: string
  age: number
  age_days: number
  tsi: number
  form: number
  stamina: number
  speed: number
  injury_days: number
  salary: number
  goalkeeper: number
  defending: number
  playmaking: number
  winger: number
  passing: number
  scoring: number
  set_pieces: number
  leadership: number
  experience: number
  loyalty: number
  market_value: number
  speciality: string | null
  last_match_rating: number | null
  transfer_listed: boolean
  country_id: number | null
  homegrown: boolean
  data_source: string
}

export interface SkillSnapshot {
  snapshot_date: string
  form: number
  stamina: number
  speed: number
  goalkeeper: number
  defending: number
  playmaking: number
  winger: number
  passing: number
  scoring: number
  set_pieces: number
  leadership: number
  experience: number
  loyalty: number
}

export function getPlayer(id: number): Promise<PlayerDetail> {
  return apiFetch<PlayerDetail>(`/api/players/${id}`)
}

export function getPlayerHistory(id: number): Promise<SkillSnapshot[]> {
  return apiFetch<SkillSnapshot[]>(`/api/players/${id}/history`)
}
